### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by marekkalnik**
Hi,

I think my refactoring is complete, however I didn't test the "following" option because it's not really clear to me how does it work. Could you give me some feedback?

Also @jtomaszewski, could you test if this is easily extensible and if you're able to adapt your compass plugin to work with it ?

Cheers,
Marek

### CODE REVIEW

**Comment by marekkalnik** [comment]
Rebase in progress.

**Comment by domoritz** [comment]
Awesome. I'll need some time to thoroughly test it but what I've seen so far is great!

**Comment by jtomaszewski** [comment]
:+1: , the code looks much more clear now.

I won't help you with testing though, as I don't use the plugin (nor the compass) anymore at this moment. 
however notify me when someone else would also need the compass functionality ;)

File: src/L.Control.Locate.js [hunk 1]
```diff
@@ -69,19 +68,167 @@ L.Control.Locate = L.Control.extend({
...
+        return this.options.markerClass(latlng, mStyle)
+    },
+
+    /**
+     * Updates the marker with current coordinates.
+     *
+     * Used by drawMarker, you can ignore it if you have overriden it.
+     */
+    updateMarker: function(latlng, mStyle) {
+        this._marker.setLatLng();
```
Comment (domoritz) [inline src/L.Control.Locate.js:207]: missing argument

**Comment by domoritz** [comment]
The idea of the follow option is that the marker follows the user's location without the need to refresh. In this mode the icon of the control has a different style (by default orange instead of blue). Would be great if you could implement it as well.

File: src/L.Control.Locate.js [hunk 2]
```diff
@@ -69,19 +68,167 @@ L.Control.Locate = L.Control.extend({
...
+
+        if (this.options.follow) {
+            this._startFollowing(map);
+        }
+    },
+
+    /**
+     * Called to stop the location engine.
+     *
+     * Override it to shutdown any functionnalities you added on start.
```
Comment (domoritz) [inline src/L.Control.Locate.js:102]: type

File: src/L.Control.Locate.js [hunk 3]
```diff
@@ -104,252 +251,206 @@ L.Control.Locate = L.Control.extend({
...
         L.DomEvent
             .on(this._link, 'click', L.DomEvent.stopPropagation)
             .on(this._link, 'click', L.DomEvent.preventDefault)
+            .on(this._link, 'dblclick', L.DomEvent.stopPropagation)
             .on(this._link, 'click', function() {
-                var shouldStop = (self._event === undefined || map.getBounds().contains(self._event.latlng)
-                    || !self.options.setView || isOutsideMapBounds());
-                if (!self.options.remainActive && (self._active && shouldStop)) {
-                    stopLocate();
+                var shouldStop = (this._event === undefined || this._map.getBounds().contains(this._event.latlng) || !this.options.setView || this._isOutsideMapBounds())
```
Comment (domoritz) [inline src/L.Control.Locate.js:261]: can you pull this into a separate function?

Comment (marekkalnik) [inline src/L.Control.Locate.js:261]: you mean the click handler ? ou just the shouldStop() line ?

Comment (domoritz) [inline src/L.Control.Locate.js:261]: The click handler

**Comment by marekkalnik** [comment]
@domoritz I'll have some spare time wednesday to integrate any comments, the beginning of this week is very busy for me :/

**Comment by domoritz** [comment]
@marekkalnik No rush. I won't have time over the next few days anyway.

**Comment by domoritz** [comment]
Hey @marekkalnik would you mind if I merged #103 before this? You will need to redo the amd work if I do.

**Comment by marekkalnik** [comment]
@domoritz no problem, I'll rebase the code

**Comment by domoritz** [comment]
Okay. It'll be tough because pretty much every line is affected. It's probably going to be easier to redo the work. I'll go ahead and test/merge #103.

**Comment by marekkalnik** [comment]
No worries, if a rewrite is needed, I'll make few smaller PR's so it is easier and faster to merge.

**Comment by domoritz** [comment]
Thanks. I went ahead and merged the AMD stuff and added npm and grunt.

**Comment by domoritz** [comment]
Any updates on this pr? I'd love to merge the refactoring. It seems great!

**Comment by marekkalnik** [comment]
@domoritz I'm creating my own company and things are a little crazy right now. it's still work in progress :/ I should be able to get to it this week.

**Comment by domoritz** [comment]
Oh wow. Good luck with your company!

**Comment by marekkalnik** [comment]
Not tested at all, but re-refactored. I should be able to test it tomorrow :)

https://github.com/marekkalnik/leaflet-locatecontrol/tree/refactoring2

**Comment by domoritz** [comment]
@marekkalnik Cool! Looking forward to it. I'm closing this pr for now.

**Pull request closed without merging**